backend/agents/agentic_rag_2.py

```diff
@@ -219,44 +235,248 @@ def _get_tools(self) -> list[QueryEngineTool]:
+
+        async def reply_func(
+            recipient: ConversableAgent,
+            messages: Optional[List[Dict]] = None,
+            sender: Optional[Agent] = None,
+            config: Optional[Any] = None,
+        ) -> Tuple[bool, Union[str, Dict, None]]:
+            logger.info(f"RAG agent received message: {messages}")
+            last_message = messages[-1].get("content")
+            response = query_knowledge_base(last_message)
+            logger.info(f"RAG agent response: {response}")
+            self._send_message("RAG Agent", response)
+            return True, response
+
+        agent.register_reply(
+            trigger=lambda sender: True,
+            reply_func=reply_func
+        )
+
+        return agent
+
+    def _create_enhancement_agent(self) -> ConversableAgent:
+        """Creates the enhancement agent"""
+        system_message = """You enhance repair instructions by:
+        1. Identifying technical terms and jargon
+        2. Adding clear explanations
+        3. Formatting for readability
+
+        Always structure your response as:
+
+        REPAIR INSTRUCTIONS:
+        [Enhanced instructions with better formatting]
+
+        TECHNICAL TERMS EXPLAINED:
+        - Term 1: Simple explanation
+        - Term 2: Simple explanation with analogy
+
+        SAFETY NOTES:
+        [Any relevant safety information]"""
+
+        agent = ConversableAgent(
+            name="enhancement_agent",
+            system_message=system_message,
+            llm_config=self.llm_config
+        )
+
+        async def reply_func(
+            recipient: ConversableAgent,
+            messages: Optional[List[Dict]] = None,
+            sender: Optional[Agent] = None,
+            config: Optional[Any] = None,
+        ) -> Tuple[bool, Union[str, Dict, None]]:
+            logger.info(f"Enhancement agent received message: {messages}")
+            flag, response = await self.enhancement_agent.a_generate_oai_reply(messages)
+            logger.info(f"Enhancement agent response: {response}")
+            self._send_message("Enhancement Agent", response)
+            return flag, response
+
+        agent.register_reply(
+            trigger=lambda sender: True,
+            reply_func=reply_func
+        )
+
+        return agent
+
+    def _is_summary(self, message: str) -> bool:
+        """Check if a message contains a complete SUMMARY block"""
+        return message.strip().startswith("SUMMARY:") and all(
+            field in message for field in ["Vehicle:", "Part/System:", "Task:"]
+        )
+
+    async def process_message(self, message: str) -> None:
+        """Main method to process user messages"""
+        logger.info("Processing message")
+        
+        # Step 1: Query Analysis
+        query_analyzer_response = await self.human_proxy.a_initiate_chat(
+            self.query_analyzer,
+            message=message,
+            clear_history=False,
+            max_turns=1
```

Comment:
just 1 iteration, is it sufficient? what if the user only provided partial info? I tried the same approach early, that's why I decided to move to group chat, I'm fine with this too tho. @beepo34

Reply:
@tony-lii thanks for the questions, I had to make a few workarounds for the back-and-forth interaction between the query_analyzer and the user to work 

- the `AgentChat` class initializes all the necessary agents and the RAG pipeline logic is handled in the member function `process_message`
- The `/chat` endpoint in `app.py` supports loading `AgentChat` instances from previous sessions using a session-id. Currently, there is no functionality to generate session-ids so all chats use `default` as the session-id. Thus each time the user sends a message, which triggers the `/chat` endpoint and calls `process_message` we are working with the SAME `AgentChat` instance.
- Thus, the `clear_history` option for `query_analyzer` is set to `False` because we want the `query_analyzer` to remember user messages from previous calls to `process_message`
- In `process_message`, when a chat is initiated with `query_analyzer`, there are 2 actions defined by the agent's system message. Both of these actions require only one reply from the `query_analyzer` which is why `max_turns=1`. Action 1 may be repeated multiple times as the `query_analyzer` requests information from the user one question at a time before moving on to Action 2 which will continue onto the rest of the RAG pipeline:
1. `query_analyzer` may prompt the user for additional information. this ends the current chat until user sends another message which triggers the `/chat` endpoint again.
2. `query_analyzer` may have all the necessary information, so it generates a `SUMMARY` which is given to the `rag_agent`

Reply:
This sounds very promising! Thank you